packages/deployer/src/compiler.ts

```diff
@@ -116,67 +150,51 @@ export class Compiler {
-
+        const fileNames = this._specifiedContracts.has(ALL_CONTRACTS_IDENTIFIER)
+            ? _.keys(this._contractSourcesIfExists)
+            : Array.from(this._specifiedContracts.values());
+        _.forEach(fileNames, fileName => {
+            this._setSourceTreeHash(fileName);
+        });
+        await Promise.all(_.map(fileNames, async fileName => this._compileContractAsync(fileName)));
         this._solcErrors.forEach(errMsg => {
             utils.consoleLog(errMsg);
         });
     }
     /**
      * Compiles contract and saves artifact to artifactsDir.
-     * @param contractBaseName Name of contract with '.sol' extension.
+     * @param fileName Name of contract with '.sol' extension.
      */
-    private async _compileContractAsync(contractBaseName: string): Promise<void> {
+    private async _compileContractAsync(fileName: string): Promise<void> {
         if (_.isUndefined(this._contractSourcesIfExists)) {
             throw new Error('Contract sources not yet initialized');
         }
+        const contractSpecificSourceData = this._contractSourceData[fileName];
+        const currentArtifact = (await this._getContractArtifactOrReturnAsync(fileName)) as ContractArtifact;
+        const sourceHash = `0x${contractSpecificSourceData.sourceHash.toString('hex')}`;
+        const sourceTreeHash = `0x${contractSpecificSourceData.sourceTreeHashIfExists.toString('hex')}`;
 
-        const source = this._contractSourcesIfExists[contractBaseName];
-        const contractName = path.basename(contractBaseName, SOLIDITY_FILE_EXTENSION);
-        const currentArtifactPath = `${this._artifactsDir}/${contractName}.json`;
-        const sourceHash = `0x${ethUtil.sha3(source).toString('hex')}`;
-        const isContractSpecified =
-            this._specifiedContracts.has(ALL_CONTRACTS_IDENTIFIER) || this._specifiedContracts.has(contractName);
-
-        let currentArtifactString: string;
-        let currentArtifact: ContractArtifact;
-        let oldNetworks: ContractNetworks;
-        let shouldCompile: boolean;
-        try {
-            const opts = {
-                encoding: 'utf8',
-            };
-            currentArtifactString = await fsWrapper.readFileAsync(currentArtifactPath, opts);
-            currentArtifact = JSON.parse(currentArtifactString);
-            oldNetworks = currentArtifact.networks;
-            const oldNetwork: ContractData = oldNetworks[this._networkId];
-            shouldCompile =
-                (_.isUndefined(oldNetwork) ||
-                    oldNetwork.keccak256 !== sourceHash ||
-                    oldNetwork.optimizer_enabled !== this._optimizerEnabled) &&
-                isContractSpecified;
-        } catch (err) {
-            shouldCompile = isContractSpecified;
-        }
-
+        const shouldCompile =
+            _.isUndefined(currentArtifact) ||
+            currentArtifact.networks[this._networkId].optimizer_enabled !== this._optimizerEnabled ||
```

Comment:
Are we absolutely sure that a key exists for `this._networkId` here? Otherwise this will give an indexing error. It would be safer to check that as well here.

Reply:
A default networkId is always passed into the constructor.